File: app/src/main/java/djordjeh/architecture/mvp/ui/tasks/TasksAdapter.java
package djordjeh.architecture.mvp.ui.tasks;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.databinding.ItemTaskBinding;

public class TasksAdapter extends RecyclerView.Adapter<TasksAdapter.TaskViewHolder> {

    interface Listener {
        void showTask(Task task);
        void saveTask(Task task);
        void deleteTask(Task task);
    }

    private Listener listener;
    private List<Task> list = new ArrayList<>();

    @NonNull
    @Override
    public TaskViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return new TaskViewHolder(ItemTaskBinding.inflate(LayoutInflater.from(viewGroup.getContext()), viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull TaskViewHolder taskViewHolder, int position) {
        taskViewHolder.bindTask(list.get(position));

        taskViewHolder.itemView.setOnClickListener(v -> {
            if (listener != null) {
                listener.showTask(list.get(taskViewHolder.getAdapterPosition()));
            }
        });

        taskViewHolder.binding.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (listener != null) {
                final Task task = list.get(taskViewHolder.getAdapterPosition());
                // breaks possible infinite loops
                if (task.isCompleted() != isChecked) {
                    task.setCompleted(isChecked);
                    listener.saveTask(task);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    boolean isEmpty() {
        return list.isEmpty();
    }

    void setList(List<Task> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    void removeTask(int index) {
        listener.deleteTask(list.get(index));
        list.remove(index);
        notifyItemRemoved(index);
    }

    void setListener(Listener listener) {
        this.listener = listener;
    }

    static class TaskViewHolder extends RecyclerView.ViewHolder {

        private ItemTaskBinding binding;

        TaskViewHolder(@NonNull ItemTaskBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bindTask(Task task) {
            binding.setTask(task);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/ui/tasks/TasksAdapter.kt
package djordjeh.architecture.mvp.ui.tasks

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.databinding.ItemTaskBinding

class TasksAdapter : RecyclerView.Adapter<TasksAdapter.TaskViewHolder>() {

    var listener: Listener? = null
    var list = mutableListOf<Task>()
    set(value) {
        field = value
        notifyDataSetChanged()
    }

    fun isEmpty() = list.isEmpty()

    interface Listener {
        fun showTask(task: Task)
        fun saveTask(task: Task)
        fun deleteTask(task: Task)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): TaskViewHolder {
        return TaskViewHolder(ItemTaskBinding.inflate(LayoutInflater.from(viewGroup.context), viewGroup, false))
    }

    override fun onBindViewHolder(taskViewHolder: TaskViewHolder, position: Int) {
        taskViewHolder.bindTask(list[position])

        taskViewHolder.itemView.setOnClickListener {
            listener?.showTask(list[taskViewHolder.adapterPosition])
        }

        taskViewHolder.binding.checkBox.setOnCheckedChangeListener { _, isChecked ->
            val task = list[taskViewHolder.adapterPosition]
            // breaks possible infinite loops
            if (task.completed != isChecked) {
                task.completed = isChecked
                listener?.saveTask(task)
            }
        }
    }

    override fun getItemCount() = list.size

    fun removeTask(index: Int) {
        listener?.deleteTask(list[index])
        list.removeAt(index)
        notifyItemRemoved(index)
    }

    class TaskViewHolder(var binding: ItemTaskBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bindTask(task: Task) {
            binding.task = task
            binding.executePendingBindings()
        }
    }
}
